core/src/telemetry/metrics.rs

```diff
@@ -428,44 +405,104 @@ static TASK_SCHED_TO_START_MS_BUCKETS: &[f64] =
-            let dname = descriptor
-                .name()
-                .strip_prefix(self.metric_prefix)
-                .unwrap_or_else(|| descriptor.name());
-            // Some recorders are just gauges
-            match dname {
-                STICKY_CACHE_SIZE_NAME | NUM_POLLERS_NAME | TASK_SLOTS_AVAILABLE_NAME => {
-                    return Some(Arc::new(last_value()))
-                }
-                _ => (),
-            }
-
-            // Other recorders will select their appropriate buckets
-            let buckets = match dname {
-                WF_E2E_LATENCY_NAME => WF_LATENCY_MS_BUCKETS,
-                WF_TASK_EXECUTION_LATENCY_NAME | WF_TASK_REPLAY_LATENCY_NAME => WF_TASK_MS_BUCKETS,
-                WF_TASK_SCHED_TO_START_LATENCY_NAME | ACT_SCHED_TO_START_LATENCY_NAME => {
-                    TASK_SCHED_TO_START_MS_BUCKETS
+fn gauge_view(metric_name: &'static str) -> opentelemetry::metrics::Result<Box<dyn View>> {
+    new_view(
+        Instrument::new().name(format!("*{metric_name}")),
+        opentelemetry_sdk::metrics::Stream::new().aggregation(Aggregation::LastValue),
+    )
+}
+
+fn histo_view(
+    metric_name: &'static str,
+    buckets: &[f64],
+) -> opentelemetry::metrics::Result<Box<dyn View>> {
+    new_view(
+        Instrument::new().name(format!("*{metric_name}")),
+        opentelemetry_sdk::metrics::Stream::new().aggregation(
+            Aggregation::ExplicitBucketHistogram {
+                boundaries: buckets.to_vec(),
+                record_min_max: true,
+            },
+        ),
+    )
+}
+
+pub(super) fn augment_meter_provider_with_views(
+    mpb: MeterProviderBuilder,
+) -> opentelemetry::metrics::Result<MeterProviderBuilder> {
+    // Some histograms are actually gauges, but we have to use histograms otherwise they forget
+    // their value between collections since we don't use callbacks.
+    Ok(mpb
+        .with_view(gauge_view(STICKY_CACHE_SIZE_NAME)?)
+        .with_view(gauge_view(NUM_POLLERS_NAME)?)
+        .with_view(gauge_view(TASK_SLOTS_AVAILABLE_NAME)?)
+        .with_view(histo_view(WF_E2E_LATENCY_NAME, WF_LATENCY_MS_BUCKETS)?)
+        .with_view(histo_view(
+            WF_TASK_EXECUTION_LATENCY_NAME,
+            WF_TASK_MS_BUCKETS,
+        )?)
+        .with_view(histo_view(WF_TASK_REPLAY_LATENCY_NAME, WF_TASK_MS_BUCKETS)?)
+        .with_view(histo_view(
+            WF_TASK_SCHED_TO_START_LATENCY_NAME,
+            TASK_SCHED_TO_START_MS_BUCKETS,
+        )?)
+        .with_view(histo_view(
+            ACT_SCHED_TO_START_LATENCY_NAME,
+            TASK_SCHED_TO_START_MS_BUCKETS,
+        )?)
+        .with_view(histo_view(ACT_EXEC_LATENCY_NAME, ACT_EXE_MS_BUCKETS)?))
+}
+
+/// OTel has no built-in synchronous Gauge. Histograms used to be able to serve that purpose, but
+/// they broke that. Lovely. So, we need to implement one by hand.
+pub(crate) struct MemoryGaugeU64 {
+    labels_to_values: Arc<RwLock<HashMap<AttributeSet, u64>>>,
+}
+
```

Comment:
I didn't immediately see test coverage for this. Worth adding some?

Reply:
It's covered by some existing tests which verify that the prometheus endpoint has the expected metrics.